### ISSUE TITLE
Embed Dagger in a CI system

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#1022]
## Overview

Dagger requires access to a complete buildkit installation. Most modern CI systems meet this requirement by offering access to a Docker engine, which can be used to bootstrap buildkit.

This means that, in theory, existing CI systems can run Dagger without requiring additional infrastructure. This is an attractive option since it makes it easier to start using Dagger, assuming you already have a CI system.

## Alternatives

There are 2 ways to run Dagger from a CI:

1. Embedded runner (described in this issue): bootstrap buildkit via Docker engine bundled with CI
2. Dedicated runner: connect to an externally managed buildkit or container runtime.

We don’t have a definitive answer to the question “which approach should I use when calling Dagger from my CI?”.

* On the one hand, Embedded has major unresolved issues, which are documented below. Because it will never be supported by 100% of CI systems, we will need to document and support Dedicated no matter what. So having to support Embedded causes feature creep.
* On the other hand, the promise of Embedded is very attractive: if you already have a good container runtime, why not use it? Dedicated is more reliable and universal, but setting it up requires more upfront work.

## Downsides of embedding

There are currently several issues with the embedding approach.

### Fragmented configuration

Many CI systems embed Docker or Docker-like capability. But each does it in a different way. Accessing that capability requires CI-specific configuration, which may or may not be documented or supported.

Solution: document and support as many CI-specific configurations as possible.

### Not universal

Some CI systems don’t bundle Docker or Docker-like capability. Embedding is not an option for those systems.

Solution: don’t make embedding the only available option for using Dagger in a CI.

### Degraded performance

Dagger relies heavily on buildkit caching for performance. For buildkit caching to work well, it requires persistent storage of some kind. That persistent storage is a basic requirement for a dedicated runner, usually configured at the containerd level, so caching “just works” without requiring any special configuration of Dagger or buildkit.

By contrast, most CIs embed a container runtime that is purely ephemeral: all data is lost in between CI runs, including the buildkit cache. This makes Dagger very slow in a naive embedded environment.

Solution: implement CI-specific workarounds when available:

* Github Actions: buildkit has recently merged the ability to import/export its cache to the proprietary storage API available in Github Actions. In addition to not being available on other CIs, this workaround is not perfect because it relies on explicit cache import/export, a buildkit feature which has its own limitations: specifically it only caches the last run, and adds an export task at the end of a run which can make total execution time longer, depending on the nature of the configuration. See #1020 

* Gitlab: no workaround available
* Jenkins: no workaround available
* CircleCI: no workaround available
* Buildkite: no workaround available

TLDR: caching issues are a major downside of the embedded model at the moment.
